base/namedtuple.jl

```diff
@@ -375,8 +375,42 @@ julia> Base.setindex(nt, "a", :a)
 (a = "a",)
 ```
 """
-function setindex(nt::NamedTuple, v, idx::Symbol)
-    merge(nt, (; idx => v))
+setindex(nt::NamedTuple, v, idx::Symbol) = merge(nt, (; idx => v))
+function setindex(nt::NamedTuple, vs, idxs::AbstractVector{Symbol})
+    merge(nt, (; [i => v for (i,v) in zip(idxs, vs)]...))
```

Comment:
There might be a little footgun hidden here, as `zip` does not check lengths:
```julia

julia> zip([:a,:c], [1,2,3])
zip([:a, :c], [1, 2, 3])

julia> collect(zip([:a,:c], [1,2,3]))
2-element Vector{Tuple{Symbol, Int64}}:
 (:a, 1)
 (:c, 2)
```
So I think there should be a length check.

Reply:
Good catch.

Reply:
Also allows duplicates, which getindex does not. Is this indended?
```
julia> _setindex((x=1, y=2, z=3), [4, 5, 6, 7], [:x, :y, :x])
(x = 6, y = 5, z = 3)

julia> (x=1, y=2, z=3)[[:x, :y, :x]]
ERROR: duplicate field name
```
Only one test https://github.com/JuliaLang/julia/pull/46453/files#diff-a4ae685ea77cc80977dc4214e7dbb9d96fed55bdb918396d3453f840da0a695bR308

Reply:
I figured this was more like the behavior of `setindex!` where a value can be overwritten multiple times.

Reply:
But get/set both allow repeats for arrays. I'm not sure whether `getindex` disallowing repeats here is a feature or a bug, either. Do you know? 

For arrays, the types must match: `setindex!(rand(10), (3, 4), 5:6)` is an error, rather than itererating the tuple, because `getindex(rand(10), 5:6)` is a vector. Is there an argument for why this method should accept & iterate anything? 
```
julia> getindex(rand(10), 5:6)
2-element Vector{Float64}:
 0.46482934129714726
 0.06656695848727301

julia> (x=1, y=2, z=3)[[:x, :y]]
(x = 1, y = 2)
```

Reply:
> But get/set both allow repeats for arrays. I'm not sure whether `getindex` disallowing repeats here is a feature or a bug, either. Do you know?

I think it's just because you can't have repeated fields in a `NamedTuple`, so `(x=1, y=2, z=3)[[:x, :y, :x]]` would have to make up a new symbol for the last `:x` in order for it to work.

> For arrays, the types must match: `setindex!(rand(10), (3, 4), 5:6)` is an error, rather than itererating the tuple, because `getindex(rand(10), 5:6)` is a vector. Is there an argument for why this method should accept & iterate anything?

I'm not completely sure what rules should carry over. You also can't index arrays with anything but integers and arrays of integers. I don't know if this means that anytime `setindex` sets multiple values the collection should be an array, or if the type should match the destination.

Reply:
Ok I guess the lack of repeats in getindex is indeed forced on you by the return type. 

The others I don't know, I just think they need careful thought.

Reply:
An alternative definition could be

```julia
function setindex(nt::NamedTuple, vs, idxs::Union{AbstractVector{Symbol},Tuple{Vararg{Symbol}}})
    merge(nt, NamedTuple{Tuple(idxs)}(vs))
end
```

which requires `vs` have a conversion to `Tuple` and matches the definition of `getindex` more.